cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -0,0 +1,325 @@
+    for (Map.Entry<String, List<Long>> entry : actionToRunRecord.entrySet()) {
+      double avgForAction = 0;
+      for (long value : entry.getValue()) {
+        avgForAction += value;
+      }
+      avgForAction /= entry.getValue().size();
+      Map temp = new HashMap<>();
+      temp.put("count", entry.getValue().size());
+      temp.put("avgRunTime", avgForAction);
+      actionToStatistic.put(entry.getKey(), temp);
+    }
+
+    for (Map.Entry<String, List<Long>> entry : actionToRunRecord.entrySet()) {
+      List<Long> runList = entry.getValue();
+      Collections.sort(runList);
+      for (double percentile : percentiles) {
+        long percentileValue = runList.get((int) ((percentile * runList.size()) / 100));
+        actionToStatistic.get(entry.getKey()).put(Double.toString(percentile), (double) percentileValue);
+      }
+    }
+
+    return new BasicStatistics(startTime, endTime, count, avgRunTime, percentileToTime,
+                                                          percentileToRunids, actionToStatistic);
+
+  }
+
+  private ImmutablePair<Map<String, Long>, Map<String, List<String>>> getPercentiles(
+    List<WorkflowRunRecord> workflowRunRecords, List<Double> percentiles) {
+    int count = workflowRunRecords.size();
+    Map<String, Long> percentileToTime = new HashMap<>();
+    Map<String, List<String>> percentileToRunids = new HashMap<>();
+    for (double i : percentiles) {
+      List<String> percentileRun = new ArrayList();
+      int percentileStart = (int) ((i * count) / 100);
+      for (int j = percentileStart; j < count; j++) {
+        percentileRun.add(workflowRunRecords.get(j).getWorkflowRunId());
+      }
+      percentileToRunids.put(Double.toString(i), percentileRun);
+      percentileToTime.put(Double.toString(i), workflowRunRecords.get(percentileStart).getTimeTaken());
+    }
+    return new ImmutablePair<>(percentileToTime, percentileToRunids);
+  }
+
+  private List<WorkflowDataset.WorkflowRunRecord> sort(List<WorkflowDataset.WorkflowRunRecord> workflowRunRecords) {
+    Collections.sort(workflowRunRecords, new Comparator<WorkflowRunRecord>() {
+      @Override
+      public int compare(WorkflowDataset.WorkflowRunRecord o1, WorkflowDataset.WorkflowRunRecord o2) {
+        return Longs.compare(o1.getTimeTaken(), o2.getTimeTaken());
+      }
+    });
+    return workflowRunRecords;
+  }
+
+  private Map<String, List<Long>> getActionRuns(List<WorkflowDataset.WorkflowRunRecord> workflowRunRecords) {
+    Map<String, List<Long>> actionToRunRecord = new HashMap<>();
+    for (WorkflowDataset.WorkflowRunRecord workflowRunRecord : workflowRunRecords) {
+      for (WorkflowDataset.ActionRuns runs : workflowRunRecord.getActionRuns()) {
+        List<Long> runList = actionToRunRecord.get(runs.getName());
+        if (runList == null) {
+          runList = new ArrayList<>();
+          actionToRunRecord.put(runs.getName(), runList);
+        } else {
```

Comment:
this can be simplified,
``
if (runList == null) {   
       actionToRunRecord.put(runs.getName(), new ArrayList<>);
     } else {
      actionToRunRecord.get(runs.getName()).add(runs.getTimeTaken));
   }
``